pintracker/optracker/operation.go

```diff
@@ -0,0 +1,192 @@
+package optracker
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	cid "github.com/ipfs/go-cid"
+
+	"github.com/ipfs/ipfs-cluster/api"
+)
+
+//go:generate stringer -type=OperationType
+
+// OperationType represents the kinds of operations that the PinTracker
+// performs and the operationTracker tracks the status of.
+type OperationType int
+
+const (
+	// OperationUnknown represents an unknown operation.
+	OperationUnknown OperationType = iota
+	// OperationPin represents a pin operation.
+	OperationPin
+	// OperationUnpin represents an unpin operation.
+	OperationUnpin
+	// OperationRemote represents an noop operation
+	OperationRemote
+)
+
+//go:generate stringer -type=Phase
+
+// Phase represents the multiple phase that an operation can be in.
+type Phase int
+
+const (
+	// PhaseError represents an error state.
+	PhaseError Phase = iota
+	// PhaseQueued represents the queued phase of an operation.
+	PhaseQueued
+	// PhaseInProgress represents the operation as in progress.
+	PhaseInProgress
+	// PhaseDone represents the operation once finished.
+	PhaseDone
+)
+
+// Operation represents an ongoing operation involving a
+// particular Cid. It provides the type and phase of operation
+// and a way to mark the operation finished (also used to cancel).
+type Operation struct {
+	ctx    context.Context
+	cancel func()
+
+	mu sync.RWMutex
+
+	opType OperationType
+	phase  Phase
+	ts     time.Time
+	pin    api.Pin
+	error  string
+}
+
+// NewOperation creates a new Operation.
+func NewOperation(ctx context.Context, pin api.Pin, typ OperationType, ph Phase) *Operation {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Operation{
+		ctx:    ctx,
+		cancel: cancel, // use *OperationTracker.Finish() instead
+
+		pin:    pin,
+		opType: typ,
+		phase:  ph,
+		ts:     time.Now(),
+		error:  "",
+	}
+}
+
+// Cid returns the Cid associated to this operation.
+func (op *Operation) Cid() *cid.Cid {
+	op.mu.RLock()
+	defer op.mu.RUnlock()
+	return op.pin.Cid
+}
+
+// Context returns the context associated to this operation.
+func (op *Operation) Context() context.Context {
+	return op.ctx
+}
+
+// Cancel will cancel the context associated to this operation.
+func (op *Operation) Cancel() {
+	op.cancel()
+}
+
+// Phase returns the Phase.
+func (op *Operation) Phase() Phase {
+	op.mu.RLock()
+	defer op.mu.RUnlock()
+	return op.phase
+}
+
+// SetPhase changes the Phase and updates the timestamp.
+func (op *Operation) SetPhase(ph Phase) {
+	op.mu.Lock()
+	defer op.mu.Unlock()
+	op.phase = ph
+	op.ts = time.Now()
+}
+
+// Error returns any error message attached to the operation.
+func (op *Operation) Error() string {
+	op.mu.RLock()
+	defer op.mu.RUnlock()
+	return op.error
+}
+
+// SetError sets an error message.
+// It updates the timestamp, but does not update the phase.
+func (op *Operation) SetError(err error) {
```

Comment:
Can this method set `phase`, `error`, and `ts` fields? I am thinking along the lines of `SetError sets the operation to an error state`, more than the current which is more of just a `setter` method.

Reply:
yep